File: lib/ParticleFilter/Map/rectangle.cpp
#include "rectangle.h"
#include "util.h"

Rectangle::Rectangle(int id, int startX, int stopX, int startY, int stopY)
    : id(id), startX(startX), stopX(stopX), startY(startY), stopY(stopY)
{
    calculateRectangleProperties();
}

Rectangle Rectangle::fromJson(const int id, const json &jsonData)
{
    return Rectangle(id,
                     jsonData["startX"],
                     jsonData["stopX"],
                     jsonData["startY"],
                     jsonData["stopY"]);
}

/// @brief Calculate rectangle properties like width and height.
void Rectangle::calculateRectangleProperties()
{
    width = stopX - startX;
    height = stopY - startY;
    diameter = sqrt(width * width + height * height);
    centerX = (startX + stopX) / 2;
    centerY = (startY + stopY) / 2;
}

/// @brief Get the width of the rectangle.
/// @return Width of the rectangle.
int Rectangle::getWidth()
{
    return this->width;
}
/// @brief Get the height of the rectangle.
/// @return Height of the rectangle.
int Rectangle::getHeight()
{
    return this->height;
}

/// @brief Get the diameter of the rectangle.
/// @return Diameter of the rectangle.
int Rectangle::getDiameter()
{
    return this->diameter;
}

/// @brief Get the center X and Y coordinate of a cell.
/// @return Center coordinates cell.
std::pair<int, int> Rectangle::getCenter() const
{
    return std::make_pair(centerX, centerY);
}

std::vector<std::pair<int, int>> Rectangle::getCoordinates() const
{
    std::vector<std::pair<int, int>> coordinates;

    coordinates.reserve(width * height);

    for (int y = startY; y <= stopY; y++)
    {
        for (int x = startX; x <= stopX; x++)
        {
            coordinates.push_back(std::pair<int, int>(x, y));
        }
    }

    return coordinates;
}

/// @brief Check if the given x/y coordinate is inside the rectangle.
/// @param x X coordinate to check.
/// @param y Y coordinate to check.
/// @return Whether or not the coordinate is inside the rectangle.
bool Rectangle::containsPoint(const int x, const int y) const
{
    return x >= startX && x <= stopX && y >= startY && y <= stopY;
}

/// @brief Check if the given x/y coordinate is inside the rectangle, excluding its border coordinates.
/// @param x X coordinate to check.
/// @param y Y coordinate to check.
/// @return Whether or not the coordinate is inside the rectangle.
bool Rectangle::containsPoint(const int x, const int y) const
{
    return x > startX && x < stopX && y > startY && y < stopY;
}

/// @brief Check whether this rectangle is inside the provided other rectangle.
/// @param other Other rectangle.
/// @return True if this rectangle is completely inside other rectangle.
bool Rectangle::isInside(Rectangle &other) const
{
    return startX >= other.startX && stopX <= other.stopX && startY >= other.startY && stopY <= other.stopY;
}

/// @brief Check if a given line intersects the coordinates of the rectangle.
/// https://www.geeksforgeeks.org/check-if-two-given-line-segments-intersect/
/// @param line Line to check for intersection.
/// @return Whether the wall is intersected by the line.
bool Rectangle::isIntersectedBy(Line line) const
{
    // Determine the four orientations needed:
    int o1 = determineOrientationThreePoints(startX, startY, stopX, stopY, line.startX, line.startY);
    int o2 = determineOrientationThreePoints(startX, startY, stopX, stopY, line.stopX, line.stopY);
    int o3 = determineOrientationThreePoints(line.startX, line.startY, line.stopX, line.stopY, startX, startY);
    int o4 = determineOrientationThreePoints(line.startX, line.startY, line.stopX, line.stopY, stopX, stopY);

    // General case:
    if (o1 != o2 && o3 != o4)
    {
        return true;
    }

    // Special case:
    if (o1 == 0 && onSegment(startX, startY, line.startX, line.startY, stopX, stopY))
    {
        return true;
    }

    if (o2 == 0 && onSegment(startX, startY, line.stopX, line.stopY, stopX, stopY))
    {
        return true;
    }

    if (o3 == 0 && onSegment(line.startX, line.startY, startX, startY, line.stopX, line.startY))
    {
        return true;
    }

    if (o4 == 0 && onSegment(line.startX, line.startY, stopX, stopY, line.stopX, line.startY))
    {
        return true;
    }

    return false;
}

/// @brief Check if the current rectangle intersects with the other rectangle at any point.
/// @param other The other rectangle.
/// @return Whether the two rectangles intersect.
bool Rectangle::isIntersectedBy(const Rectangle &other, bool ignoreEdge) const
{
    if (ignoreEdge)
    {
        return startX < other.stopX && stopX > other.startX && startY < other.stopY && stopY > other.startY;
    }

    return startX <= other.stopX && stopX >= other.startX && startY <= other.stopY && stopY >= other.startY;
}

/// @brief Update the stop coordinate of the rectangle.
/// @param newStopX The new stop coordinate of the X-axis.
void Rectangle::updateStopX(const int newStopX)
{
    this->stopX = stopX;

    calculateRectangleProperties();
}

/// @brief Update the stop coordinate of the rectangle.
/// @param newStopY The new stop coordinate of the Y-axis.
void Rectangle::updateStopY(const int newStopY)
{
    this->stopY = stopY;

    calculateRectangleProperties();
}

File: Python/LocalizationTable/Map/rectangle.py
import math
from typing import List, Tuple
from .Line import Line


def on_segment(p1x, p1y, p2x, p2y, p3x, p3y):
    return (p2x <= max(p1x, p3x) and p2x >= min(p1x, p3x)) and (p2y <= max(p1y, p3y) and p2y >= min(p1y, p3y))


def determine_orientation_three_points(p1x, p1y, p2x, p2y, p3x, p3y):
    val = (p2y - p1y) * (p3x - p2x) - (p2x - p1x) * (p3y - p2y)

    return 0 if val == 0 else (1 if val > 0 else 2)


class Rectangle(object):
    def __init__(self, id: int, start_x: int, stop_x: int, start_y: int, stop_y: int):
        self.center_x = None
        self.center_y = None
        self.diameter = None
        self.height = None
        self.width = None
        self.id = id
        self.start_x = start_x
        self.stop_x = stop_x
        self.start_y = start_y
        self.stop_y = stop_y
        self.calculate_rectangle_properties()

    @staticmethod
    def from_json(id: int, jsonData: dict) -> "Rectangle":
        return Rectangle(id, jsonData["startX"], jsonData["stopX"], jsonData["startY"], jsonData["stopY"])

    def calculate_rectangle_properties(self) -> None:
        self.width = self.stop_x - self.start_x
        self.height = self.stop_y - self.start_y
        self.diameter = math.sqrt(self.width * self.width + self.height * self.height)
        self.center_x = (self.start_x + self.stop_x) // 2
        self.center_y = (self.start_y + self.stop_y) // 2

    def get_width(self) -> int:
        return self.width

    def get_height(self) -> int:
        return self.height

    def get_diameter(self) -> int:
        return self.diameter

    def get_center(self) -> Tuple[int, int]:
        return self.center_x, self.center_y

    def get_coordinates(self) -> List[Tuple[int, int]]:
        coordinates = []

        for y in range(self.start_y, self.stop_y + 1):
            for x in range(self.start_x, self.stop_x + 1):
                coordinates.append((x, y))

        return coordinates

    def contains_point(self, x: int, y: int) -> bool:
        return self.start_x <= x <= self.stop_x and self.start_y <= y <= self.stop_y

    def contains_point_excluding_border(self, x: int, y: int) -> bool:
        return self.start_x < x < self.stop_x and self.start_y < y < self.stop_y

    def is_inside(self, other: "Rectangle") -> bool:
        return self.start_x >= other.start_x and self.stop_x <= other.stop_x and self.start_y >= other.start_y and self.stop_y <= other.stop_y

    def is_intersected_by(self, line: Line):
        # Determine the four orientations needed
        o1 = determine_orientation_three_points(self.start_x, self.start_y, self.stop_x, self.stop_y, line.start_x,
                                                     line.start_y)
        o2 = determine_orientation_three_points(self.start_x, self.start_y, self.stop_x, self.stop_y, line.stop_x,
                                                     line.stop_y)
        o3 = determine_orientation_three_points(line.start_x, line.start_y, line.stop_x, line.stop_y, self.start_x,
                                                     self.start_y)
        o4 = determine_orientation_three_points(line.start_x, line.start_y, line.stop_x, line.stop_y, self.stop_x,
                                                     self.stop_y)

        # General case
        if o1 != o2 and o3 != o4:
            return True

        # Special cases
        if o1 == 0 and on_segment(self.start_x, self.start_y, line.start_x, line.start_y, self.stop_x,
                                       self.stop_y):
            return True

        if o2 == 0 and on_segment(self.start_x, self.start_y, line.stop_x, line.stop_y, self.stop_x, self.stop_y):
            return True

        if o3 == 0 and on_segment(line.start_x, line.start_y, self.start_x, self.start_y, line.stop_x,
                                       line.stop_y):
            return True

        if o4 == 0 and on_segment(line.start_x, line.start_y, self.stop_x, self.stop_y, line.stop_x, line.stop_y):
            return True

        return False

    def is_intersected_by_rect(self, other: "Rectangle", ignore_edge: bool = False) -> bool:
        if ignore_edge:
            return self.start_x < other.stop_x and self.stop_x > other.start_x and self.start_y < other.stopY and self.stop_y > other.start_y

        return self.start_x <= other.stop_x and self.stop_x >= other.start_x and self.start_y <= other.stopY and self.stop_y >= other.start_y

    def update_stop_x(self, new_stop_x: int) -> None:
        self.stop_x = new_stop_x
        self.calculate_rectangle_properties()

    def update_stop_y(self, new_stop_y: int) -> None:
        self.stop_y = new_stop_y
        self.calculate_rectangle_properties()
